# dataquest.io - Python Programming
## Guided project - explore US Births

This notepad analyses the birth rates from the US between 1993 and 2014 using the CDC and SSA data sets that are available [here](https://github.com/fivethirtyeight/data/tree/master/births).

The following functions are defined:
* `read_csv(file_name)`
* `calc_counts(data, column)`
* `calc_max_dict_value(dictionary)`
* `calc_min_dict_value(dictionary)`
* `calc_counts_with_condition(data,condition_col, condition_val, group_by_col,value_col=4)`
* `calc_diff_counts(data)`
* `calc_diff_counts_with_condition(data,condition_col, condition_val, group_by_col,value_col=4)`
* `create_date_key(data)`
* `create_dictionary_with_date_key(data)`
* `combine_dictionaries(dict_1, dict_2, col_to_merge)`


In [42]:
def read_csv(file_name):
    file = open(file_name).read()
    data = file.split("\n")
    end_index = len(data)
    string_list = data[1:end_index]
    
    final_list = []
    
    for row in string_list:
        int_fields = []
        string_fields = row.split(",")
        
        for field in string_fields:
            field_value = int(field)
            int_fields.append(field_value)
            
        final_list.append(int_fields)
    
    return final_list  

In [43]:
cdc_list = read_csv("US_births_1994-2003_CDC_NCHS.csv")
cdc_list[0:10]

[[1994, 1, 1, 6, 8096],
 [1994, 1, 2, 7, 7772],
 [1994, 1, 3, 1, 10142],
 [1994, 1, 4, 2, 11248],
 [1994, 1, 5, 3, 11053],
 [1994, 1, 6, 4, 11406],
 [1994, 1, 7, 5, 11251],
 [1994, 1, 8, 6, 8653],
 [1994, 1, 9, 7, 7910],
 [1994, 1, 10, 1, 10498]]

In [44]:
def calc_counts(data, column):
    final_list = {}
    
    for data_row in data:
        column_value = data_row[column]
        births = data_row[4]
        
        if column_value in final_list:
            final_list[column_value] = final_list[column_value] + births
        else:
            final_list[column_value] = births
        
    return final_list

In [45]:
cdc_year_births = calc_counts(cdc_list, 0)
cdc_month_births = calc_counts(cdc_list, 1)
cdc_dom_births = calc_counts(cdc_list, 2)
cdc_dow_births = calc_counts(cdc_list, 3)

In [46]:
def calc_max_dict_value(dictionary):
      return max(dictionary.values())
    
def calc_min_dict_value(dictionary):
      return min(dictionary.values())

### calc_counts_with_condition

Returns a dictionary from a list of lists that contains the summed values of elements in rows that meet a specified condition.
* `condition_col` - the column to check the value of as part of the condition
* `condition_val` - the value the condition column value needs to equal to be included in the returned dictionary
* `group_by_col` - the column to use as the 'key' for each group of summed values (i.e. 0 = year, 1 = month)
* `value_col` - the column containing the value to be used when summing the rows that meet the condition (default = births [4])


In [47]:
def calc_counts_with_condition(data,condition_col, condition_val, group_by_col,value_col=4):
    grouped_list = {}
    
    for row in data:
        key = row[group_by_col]
        value = row[value_col]
        condition = row[condition_col]
        
        if condition == condition_val:
            if key in grouped_list:
                grouped_list[key] = grouped_list[key] + value
            else:
                grouped_list[key] = value
            
    return grouped_list

## calc_diff_counts
Calculates the difference between values within a dictionary from one element to the next. Elements are ordered by the dictionary key 

In [48]:
def calc_diff_counts(data):
    diff_list = {}
    
    sorted_list = sorted(data)
    prev_value = 0
    
    for row in sorted_list:
        diff_list[row] = data[row] - prev_value
        prev_value = data[row]
        
    return diff_list
            

## calc_diff_counts_with_condition
Combines the retrieval of values that meet a given condition using the `calc_counts_with_condition` function, with the `calc_diff_counts` function

In [49]:
def calc_diff_counts_with_condition(data,condition_col, condition_val, group_by_col,value_col=4):
    grouped_list = calc_counts_for_condition(cdc_list,condition_col,condition_val,group_by_col,value_col)
    diff_list = calc_diff_counts(grouped_list)
    
    return diff_list

Extracts the difference in all births on Saturdays (specified via `condition_column` & `condition_value`) from one year (specified by `group_by_column`) to the next

In [50]:
condition_column = 3 # day of week
condition_value = 6 # saturday
group_by_column = 0 # year

diff_list = calc_diff_counts_with_condition(combined_list,condition_column,condition_value,group_by_column)
diff_list

NameError: name 'combined_list' is not defined

# Merge two datasets with same structure

The following cells merge to data sets (SSA (2000 to 2014) and CDC (1994 - 2003)) into  a single list.

Where the two data sets overlap on a date, the mean of the values within both data sets is uses in place of the original values (decimal places removed by converting float value to int)

Each data set is converted to a dictionary with the key being composed of the unique date for each data item, and the value being the data item, alowing matching rows to be found using a straightforward check that a key in one list exists in the other, i.e.

[1994, 1, 7, 5, 11251] becomes { **'1994.1.7'**: [1994, 1, 7, 5, 11251]}

In [ ]:
ssa_list = read_csv("US_births_2000-2014_SSA.csv")

In [ ]:
def create_date_key(data):
    return str(data[0]) + '.' + str(data[1]) + '.' + str(data[2])

def create_dictionary_with_date_key(data):
    dictionary = {}
    for row in data:
        key = create_date_key(row)
        dictionary[key] = row
    return dictionary

date_keyed_ssa_dictionary = create_dictionary_with_date_key(ssa_list)
date_keyed_cdc_dictionary = create_dictionary_with_date_key(cdc_list)


In [ ]:
def merge_dictionaries(dict_1, dict_2, col_to_merge=4):
       
    merged_dict = dict_1
    
    for row in dict_2:
        if row in dict_1:
            value_1 = dict_1[row][col_to_merge]
            value_2 = dict_2[row][col_to_merge]
            mean = int((value_1 + value_2) / 2)
            
            merged_dict[row][col_to_merge] = mean
        else:
            merged_dict[row] = dict_2[row]
           
    return merged_dict


In [51]:
merged_dictionaries = merge_dictionaries(date_keyed_cdc_dictionary,date_keyed_ssa_dictionary)
merged_list = [ v for v in merged_dictionaries.values()]

merged_list[0:10]

[[1997, 7, 24, 4, 11988],
 [2005, 9, 3, 6, 9128],
 [2006, 4, 8, 6, 8333],
 [2000, 8, 13, 7, 8312],
 [1996, 7, 16, 2, 12619],
 [2012, 9, 22, 6, 8899],
 [2009, 10, 9, 5, 12706],
 [1996, 5, 7, 2, 11567],
 [2000, 10, 31, 2, 11370],
 [2012, 3, 6, 2, 12076]]